src/codegen/encoding/buffer_validator.rs

```diff
@@ -0,0 +1,201 @@
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
+use crate::codegen::vartable::Vartable;
+use crate::codegen::Expression;
+use crate::sema::ast::{Namespace, Type};
+use num_bigint::BigInt;
+use num_traits::Zero;
+use solang_parser::pt::Loc;
+use std::ops::AddAssign;
+
+/// When we are decoding serialized data from a bytes array, we must constantly verify if
+/// we are not reading past its ending. This struct helps us decrease the number of checks we do,
+/// by merging checks when we can determine the size of what to read beforehand.
+pub(super) struct BufferValidator<'a> {
+    /// Saves the codegen::Expression that contains the buffer length.
+    pub(super) buffer_length: Expression,
+    /// The types we are supposed to decode
+    pub(super) types: &'a [Type],
+    /// The argument whose size has already been accounted for when verifying the buffer
+    pub(super) verified_until: i32,
+    /// The argument we are analysing presently
+    pub(super) current_arg: usize,
```

Comment:
`verified_until` and `current_arg` are different types. I think this is so because you want to assign -1 to `verfied_until`. As as result there are a bunch `as i32` casts in the code, which are ugly.

How about making `verified_until` an `Option<usize>` or special casing `fn initialize_validation()` so that it does not need to set `-1`?

Reply:
> special casing fn initialize_validation() so that it does not need to set -1?

This was my first attempt, however, when the first argument to decode is a dynamic array, the validator must still be aware that it has no been verified, even after calling `initialize_validation`. I'll implement your `Option<usize>` suggestion.